bionic/code_analyzer.py

```diff
@@ -0,0 +1,134 @@
+        self._analyzed_bytearray = bytearray()
+
+    def get_analyzed_bytes(self):
+        return self._analyzed_bytearray
+
+    def analyze(self, obj):
+        obj_id = get_obj_id(obj)
+
+        # If already hashed, return the hash.
+        if obj_id in self._id_to_hash:
+            return self._id_to_hash[obj_id]
+
+        # If the obj is already hashing, break the circular ref and return early.
+        if obj_id in self._ids_being_hashed:
+            return get_obj_name(obj, DEFAULT_NAME)
+        self._ids_being_hashed.add(obj_id)
+
+        # Otherwise, hash the object.
+        obj_hash = self._analyze(obj)
+        self._id_to_hash[obj_id] = obj_hash
+        self._ids_being_hashed.remove(obj_id)
+
+    def _update_bytes(self, data):
+        self._analyzed_bytearray += data
+
+    def _analyze(self, obj):
+        if isinstance(obj, bytes) or isinstance(obj, bytearray):
+            self._update_bytes(obj)
+
+        elif obj is None or isinstance(obj, (int, float, str, bool)):
+            json_bytes = json.dumps(obj).encode()
+            self._update_bytes(json_bytes)
+
+        elif isinstance(obj, (list, tuple)):
+            for elem in obj:
+                self.analyze(elem)
+
+        elif isinstance(obj, dict):
+            for key, elem in obj.items():
+                self.analyze(key)
+                self.analyze(elem)
+
+        elif inspect.isroutine(obj):
+            if obj.__defaults__ is not None:
+                for d in obj.__defaults__:
+                    self.analyze(d)
+
+            self._analyze_code(obj.__code__)
+
+        elif inspect.iscode(obj):
+            self._analyze_code(obj)
+
+        elif inspect.isclass(obj):
+            # TODO: Add support to analyze class objects.
+            self.analyze(obj.__name__)
+
+        else:
+            message = oneline(
+                f"""
+                Found complex object {obj!r} of type {type(obj)!r} while analyzing
+                code for caching, probably as a global variable or a function
+                default. Any changes to its value won't be detected by Bionic,
+                which may result in Bionic using stale cache values. Consider
+                making this value a Bionic entity instead.
+
+                Check https://bionic.readthedocs.io/en/stable/warnings.html#avoid-global-state
+                for more information.
+                """
+            )
+            warnings.warn(message)
+
+    def _analyze_code(self, code):
+        # TODO: Find references for the code and analyze references.
+        self.analyze(code.co_code)
+
+        # Hash constants used in the code object, but ignore names of lambdas.
+        consts = [
+            const
+            for const in code.co_consts
+            if not isinstance(const, str) or not const.endswith(".<lambda>")
+        ]
+        self.analyze(consts)
+
+
+def get_obj_id(obj):
+    if obj is None or isinstance(obj, (int, float, str, bool)):
```

Comment:
Why the special case here? Doesn't this mean that two objects could share the same "id"?

Also, if we're going to return something that's different from `id`, we should have a different name.

Reply:
Streamlit calls it key, I guess we can call it the same.

Reply:
I feel so stupid to miss this. I can remove the special case. I thought it would be nice to skip encoding the same integer or string multiple times, but that should be quick anyway.